cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/AbstractStorageProviderNamespaceAdmin.java

```diff
@@ -139,54 +140,89 @@ private void createLocation(NamespaceMeta namespaceMeta) throws IOException {
+      createdStreams = createNamespaceDir(streamsLoc, "streams", namespaceId);
+      createNamespaceDir(deletedLoc, "deleted streams", namespaceId);
+
+      // then set all these directories to be owned and group writable by the namespace group as follows:
+      // if a group name is configured, then that group; otherwise the same group as the namespace home dir
       if (SecurityUtil.isKerberosEnabled(cConf)) {
-        // the data dir should have the group from the namespace config if present, or the same group as the home dir
-        String dataGroup = configuredGroupName != null ? configuredGroupName : namespaceHome.getGroup();
-        namespaceData.setGroup(dataGroup);
-        // set the permissions to rwx for group, if a group name was configured for the namespace
-        if (configuredGroupName != null) {
-          String permissions = namespaceData.getPermissions();
-          namespaceData.setPermissions(permissions.substring(0, 3) + "rwx" + permissions.substring(6));
+        String groupToSet = configuredGroupName != null ? configuredGroupName : namespaceHome.getGroup();
+        for (Location loc : new Location[] { dataLoc, tempLoc, streamsLoc, deletedLoc }) {
+          loc.setGroup(groupToSet);
+          // set the permissions to rwx for group, if a group name was configured for the namespace
+          if (configuredGroupName != null) {
+            String permissions = loc.getPermissions();
+            loc.setPermissions(permissions.substring(0, 3) + "rwx" + permissions.substring(6));
+          }
         }
       }
     } catch (Throwable t) {
-      try {
-        if (createdHome) {
-          namespaceHome.delete(true);
-        } else if (createdData) {
-          namespaceData.delete(true);
+      if (createdHome) {
+        deleteDirSilently(namespaceHome, t, "home", namespaceMeta.getNamespaceId());
+      } else {
+        if (createdData) {
+          deleteDirSilently(dataLoc, t, "data", namespaceMeta.getNamespaceId());
+        }
+        if (createdTemp) {
+          deleteDirSilently(tempLoc, t, "temp", namespaceMeta.getNamespaceId());
+        }
+        if (createdStreams) {
+          deleteDirSilently(streamsLoc, t, "streams", namespaceMeta.getNamespaceId());
         }
-      } catch (Throwable t1) {
-        LOG.warn("Error while cleaning up home directory '%s' for namespace '%s'",
-                 namespaceHome, namespaceMeta.getNamespaceId());
-        t.addSuppressed(t1);
       }
-      Throwables.propagateIfInstanceOf(t, IOException.class);
-      throw Throwables.propagate(t);
+      throw t;
+    }
+  }
+
+  // create the location and return true; throw IOException if the location could not be created
+  private boolean createNamespaceDir(Location location, String what, NamespaceId namespace) throws IOException {
+    try {
+      if (!location.mkdirs()) {
+        throw new IOException(String.format("Error while creating %s directory '%s' for namespace '%s': %s",
```

Comment:
This will be caught by the outer catch and be needlessly nested. Refactor to avoid nesting?